src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -35,102 +32,264 @@ export saturation_excess
+All `ThermodynamicState`'s have access to functions to
+compute all other thermodynamic properties.
 """
-function soundspeed_air(T, q_tot=0, q_liq=0, q_ice=0)
+abstract type ThermodynamicState{DT} end
 
-    _γ   = cp_m(q_tot, q_liq, q_ice)/cv_m(q_tot, q_liq, q_ice)
-    _R_m = gas_constant_air(q_tot, q_liq, q_ice)
-    return sqrt(_γ * _R_m * T)
+"""
+    InternalEnergySHumEquil{DT} <: ThermodynamicState
+
+A thermodynamic state initialized by
+
+ - `e_int` internal energy
+ - `q_tot` total specific humidity
+ - `ρ` density
+
+assuming thermodynamic equilibrium (therefore, saturation
+adjustment is needed).
+"""
+struct InternalEnergySHumEquil{DT} <: ThermodynamicState{DT}
+    e_int::DT
+    q_tot::DT
+    ρ::DT
+    T::DT
+end
+function InternalEnergySHumEquil(e_int, q_tot, ρ)
+    return InternalEnergySHumEquil(e_int, q_tot, ρ, saturation_adjustment(e_int, ρ, q_tot))
+end
+
+"""
+    InternalEnergySHumNonEquil{DT} <: ThermodynamicState
+
+A thermodynamic state initialized by
+
+ - `e_int` internal energy
+ - `q_tot` total specific humidity
+ - `q_liq` specific humidity of liquid
+ - `q_ice` specific humidity of ice
+ - `ρ` density
 
+assuming thermodynamic non-equilibrium (therefore,
+temperature can be computed directly).
+"""
+struct InternalEnergySHumNonEquil{DT} <: ThermodynamicState{DT}
+    e_int::DT
+    q_tot::DT
+    q_liq::DT
+    q_ice::DT
+    ρ::DT
 end
 
 """
+    LiquidPotTempSHumEquil{DT} <: ThermodynamicState{DT}
+
+A thermodynamic state initialized by
+ - `θ_liq` liquid potential temperature
+ - `q_tot` total specific humidity
+ - `ρ` density
+ - `p` pressure
+
+assuming thermodynamic equilibrium (therefore, saturation
+adjustment is needed).
+"""
+struct LiquidPotTempSHumEquil{DT} <: ThermodynamicState{DT}
+    θ_liq::DT
+    q_tot::DT
+    ρ::DT
+    p::DT
+    T::DT
+end
+function LiquidPotTempSHumEquil(θ_liq, q_tot, ρ, p)
+    T = saturation_adjustment_q_t_θ_l(θ_liq, q_tot, ρ, p)
+    return LiquidPotTempSHumEquil(θ_liq, q_tot, ρ, p, T)
+end
+
+"""
+    gas_constant_air(ts::ThermodynamicState)
     gas_constant_air([q_tot=0, q_liq=0, q_ice=0])
 
-Return the specific gas constant of moist air, given the total specific
-humidity `q_tot`, and, optionally, the liquid specific humidity `q_liq`,
-and the ice specific humidity `q_ice`. When no input argument is given, it
-returns the specific gas constant of dry air.
+Return the specific gas constant of moist air, given
+a thermodynamic state `ts` or, optionally,
+
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
+
+Without the specific humidity arguments, it the results
+are that of dry air.
 """
-function gas_constant_air(q_tot=0, q_liq=0, q_ice=0)
+function gas_constant_air(q_tot::DT=0.0, q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
 
-    return R_d * ( 1 +  (molmass_ratio - 1)*q_tot - molmass_ratio*(q_liq + q_ice) )
+    return DT(R_d) * ( 1 +  (DT(molmass_ratio) - 1)*q_tot - DT(molmass_ratio)*(q_liq + q_ice) )
```

Comment:
Would be good to replace `molmass_ratio` here (and in PlanetParameters and SurfaceFluxes) by `\epsilon_dv` as in atmosphere design doc.

Reply:
I like this idea, but maybe this can be a separate PR? (I think this is used in the, and I’d like to keep changes local)

Reply:
I like this idea, but maybe this can be a separate PR? (I think this is used in the, and I’d like to keep changes local)

Reply:
Sure, a separate PR is fine.